source/fe/mapping_q.cc

```diff
@@ -88,18 +89,33 @@ template <int dim, int spacedim>
 MappingQ<dim,spacedim>::MappingQ (const MappingQ<dim,spacedim> &mapping)
   :
   polynomial_degree (mapping.polynomial_degree),
-  use_mapping_q_on_all_cells (mapping.use_mapping_q_on_all_cells),
-  // clone the Q1 mapping for use on interior cells (if necessary)
-  // or to create a good initial guess in transform_real_to_unit_cell()
-  q1_mapping (dynamic_cast<MappingQGeneric<dim,spacedim>*>(mapping.q1_mapping->clone())),
-  // create a Q_p mapping; if p=1, simply share the Q_1 mapping already
-  // created via the shared_ptr objects
-  qp_mapping (this->polynomial_degree>1
-              ?
-              std::shared_ptr<const MappingQGeneric<dim,spacedim> >(dynamic_cast<MappingQGeneric<dim,spacedim>*>(mapping.qp_mapping->clone()))
-              :
-              q1_mapping)
-{}
+  use_mapping_q_on_all_cells (mapping.use_mapping_q_on_all_cells)
+{
+  // Note that we really do have to use clone() here, since mapping.q1_mapping
+  // may be MappingQ1Eulerian and mapping.qp_mapping may be MappingQEulerian.
+  std::unique_ptr<Mapping<dim,spacedim>> other_q1_map = mapping.q1_mapping->clone();
+  const auto q1_map = dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_q1_map.get());
+  Assert(q1_map != nullptr, ExcInternalError());
+  Assert(q1_map->get_degree() == 1, ExcInternalError());
+  q1_mapping.reset(q1_map);
+  // avoid double delete
+  other_q1_map.release();
```

Comment:
Calling `release()` without assigning always looks really dangerous to me.
I don't see a reason why we can't do
```
std::unique_ptr<Mapping<dim,spacedim>> other_q1_map = mapping.q1_mapping->clone();
q1_mapping.reset(dynamic_cast<const MappingQGeneric<dim,spacedim> *>(other_q1_map.release()));
Assert(q1_mapping, ExcInternalError());
Assert(q1_mapping->get_degree() == 1, ExcInternalError());
```
directly avoiding the unassigned `release()` call and the local variable `q1_map`.

Reply:
Thats nicer than what I did. I'll update it.